File: search/binary_search/iterative/go/binary_search.go
package main

import (
	"fmt"
)

func binary_search(values_slice []int, value int) bool {
	start := 0
	end := len(values_slice) - 1

	for start <= end {
		mid := (end + start) / 2

		if values_slice[mid] == value {
			return true
		} else if values_slice[mid] > value {
			end = mid - 1
		} else {
			start = mid + 1
		}
	}

	return false
}

func main() {
	values := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
	fmt.Println(binary_search(values, 8))
}

File: search/binary_search/iterative/python/binary_search.py
def binary_search(lst, value):
	start = 0
	end = len(lst) - 1

	while start <= end:
		mid = (end + start) // 2
		if lst[mid] == value:
			return True
		elif lst[mid] < value:
			start = mid + 1
		else:
			end = mid - 1
	return False

def test_binary_search(lst, value, resp):
	result = binary_search(lst, value)
	assert  result == resp
	if result:
		print lst, ",has value: ", value
	else:
		print lst, ",hasn't value: ", value


test_binary_search([1,2,3,4,5,6,7,8,9], 9, True)
test_binary_search([1,2,3,4,5,6,7,8,9], 2, True)
test_binary_search([1,2,3,4,5,6,7,8,9], 11, False)
test_binary_search([1,2,3,4,5,6,7,8,9], 10, False)

